python/cuspatial/cuspatial/geometry/geoseries.py

```diff
@@ -0,0 +1,457 @@
+            elif (
+                self.source.types[ml_index] == "l" and preceding_ml_count == 0
+            ):
+                preceding_line_count = preceding_line_count + 1
+            ml_index = ml_index - 1
+        preceding_multis = preceding_ml_count
+        if preceding_multis > 0:
+            multi_end = self.source._lines.mlines[preceding_multis * 2 - 1]
+            item_start = multi_end + preceding_line_count
+        else:
+            item_start = preceding_line_count
+        item_length = self.source.lengths[self.index]
+        item_end = item_length + item_start
+        item_source = self.source._lines
+        result = item_source[item_start:item_end]
+        return LineString(
+            result.to_array().reshape(2 * (item_start - item_end), 2)
+        )
+
+
+class cuMultiLineString(cuGeometry):
+    def to_shapely(self):
+        item_type = self.source.types[self.index]
+        index = (
+            pd.Series(self.source.types[0 : self.index]) == item_type
+        ).sum()
+        line_indices = slice(
+            self.source._lines.mlines[index * 2],
+            self.source._lines.mlines[index * 2 + 1],
+        )
+        lines = []
+        for i in range(line_indices.start, line_indices.stop, 1):
+            line = self.source._lines[i].to_array()
+            lines.append(LineString(line.reshape(int(len(line) / 2), 2)))
+        return MultiLineString(lines)
+
+
+class cuPolygon(cuGeometry):
+    def to_shapely(self):
+        mp_index = self.index - 1
+        preceding_poly_count = 0
+        preceding_mp_count = 0
+        while mp_index >= 0:
+            if self.source.types[mp_index] == "mpoly":
+                preceding_mp_count = preceding_mp_count + 1
+            elif (
+                self.source.types[mp_index] == "poly"
+                and preceding_mp_count == 0
+            ):
+                preceding_poly_count = preceding_poly_count + 1
+            mp_index = mp_index - 1
+        preceding_multis = preceding_mp_count
+        multi_index = (
+            self.source._polygons.mpolys[preceding_multis * 2 - 1]
+            if preceding_multis > 0
+            else 0
+        )
+        preceding_polys = preceding_poly_count
+        ring_start = self.source._polygons.polys[multi_index + preceding_polys]
+        ring_end = self.source._polygons.polys[
+            multi_index + preceding_polys + 1
+        ]
+        rings = self.source._polygons.rings
+        exterior_slice = slice(rings[ring_start], rings[ring_start + 1])
+        exterior = self.source._polygons.xy[exterior_slice]
+        interiors = []
+        for interior in range(ring_end - (ring_start + 1)) + (ring_start + 1):
+            interior_slice = slice(rings[interior], rings[interior + 1])
+            interiors.append(
+                self.source._polygons.xy[interior_slice]
+                .to_array()
+                .reshape(int((rings[interior + 1] - rings[interior]) / 2), 2)
+            )
```

Comment:
nit: Could turn this into a List comprehension

Reply:
:heavy_check_mark: